packages/cojson/src/coValueCore/SessionMap.ts

```diff
@@ -0,0 +1,231 @@
+      };
+      this.sessions.set(sessionID, sessionLog);
+    }
+
+    return sessionLog;
+  }
+
+  addTransaction(
+    sessionID: SessionID,
+    signerID: SignerID,
+    newTransactions: Transaction[],
+    newSignature: Signature,
+    skipVerify: boolean = false,
+  ): Result<true, TryAddTransactionsError> {
+    const sessionLog = this.getOrCreateSessionLog(sessionID, signerID);
+
+    try {
+      // TODO: we don't use the return value here
+      sessionLog.impl.tryAdd(
+        newTransactions.map((tx) => stableStringify(tx)),
+        newSignature,
+        skipVerify,
+      );
+
+      this.addTransactionsToJsLog(sessionLog, newTransactions, newSignature);
+
+      return ok(true as const);
+    } catch (e) {
+      return err({
+        type: "InvalidSignature",
+        id: this.id,
+        sessionID,
+        newSignature,
+        signerID,
+        error: e as Error,
+      } satisfies TryAddTransactionsError);
+    }
+  }
+
+  makeNewTransaction(
+    sessionID: SessionID,
+    signerAgent: ControlledAccountOrAgent,
+    changes: JsonValue[],
+    privacy:
+      | { type: "private"; keyID: KeyID; keySecret: KeySecret }
+      | { type: "trusting" },
+  ): { signature: Signature; transaction: Transaction } {
+    const sessionLog = this.getOrCreateSessionLog(
+      sessionID,
+      signerAgent.currentSignerID(),
+    );
+
+    const madeAt = Date.now();
+
+    let signatureAndTxJson: string;
+
+    if (privacy.type === "private") {
+      signatureAndTxJson = sessionLog.impl.addNewPrivateTransaction(
+        stableStringify(changes),
+        signerAgent.currentSignerSecret(),
+        privacy.keySecret,
+        privacy.keyID,
+        madeAt,
+      ) as Signature;
+    } else {
+      signatureAndTxJson = sessionLog.impl.addNewTrustingTransaction(
+        stableStringify(changes),
+        signerAgent.currentSignerSecret(),
+        madeAt,
+      ) as Signature;
+    }
+
+    // TODO: we don't use the hash here
+    const { signature, transaction } = JSON.parse(signatureAndTxJson);
+
+    this.addTransactionsToJsLog(sessionLog, [transaction], signature);
+
+    return { signature, transaction };
+  }
+
+  private addTransactionsToJsLog(
+    sessionLog: SessionLog,
+    newTransactions: Transaction[],
+    signature: Signature,
+  ) {
+    for (const tx of newTransactions) {
+      sessionLog.transactions.push(tx);
+    }
+    sessionLog.lastSignature = signature;
+
+    sessionLog.txSizeSinceLastInbetweenSignature += newTransactions.reduce(
+      (sum, tx) =>
+        sum +
+        (tx.privacy === "private"
+          ? tx.encryptedChanges.length
+          : tx.changes.length),
+      0,
+    );
+
+    if (
+      sessionLog.txSizeSinceLastInbetweenSignature > MAX_RECOMMENDED_TX_SIZE
+    ) {
+      sessionLog.signatureAfter[sessionLog.transactions.length - 1] = signature;
+      sessionLog.txSizeSinceLastInbetweenSignature = 0;
+    }
+  }
+
+  // TODO: Shall we get rid of this?
+  testExpectedHashAfter(
```

Comment:
This method is used only on some tests.

I think it would be good to get rid of it, if you agree I can do the work directly on main.